cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/GoalOptimizer.java

```diff
@@ -117,6 +118,37 @@ public GoalOptimizer(KafkaCruiseControlConfig config,
     _proposalComputationTimer = dropwizardMetricRegistry.timer(MetricRegistry.name("GoalOptimizer", "proposal-computation-timer"));
   }
 
+  /**
+   * Generate unique goal priority for each of proposal precomupting thread
+   */
+  private void populateGoalByPriorityForPrecomputing() {
+    Set<List<Goal>> shuffledGoalSet = new HashSet<>();
+    //Need at least one computing thread
+    for (int i = 0; i < numProposalComputingThreads(); i++) {
+      if (i == 0) {
+        //Guarantee one thread is working on the original goal priority
+        _goalByPriorityForPrecomputing.add(_goalsByPriority);
+        shuffledGoalSet.add(new ArrayList(_goalsByPriority.values()));
+      } else {
+        List<Goal> shuffledGoal = new ArrayList(_goalsByPriority.values());
+        //Ensure the shuffled goal is unique
+        do {
+          Collections.shuffle(shuffledGoal);
+        } while (! shuffledGoalSet.add(shuffledGoal));
```

Comment:
This loop might take undeterministic time to complete -- potentially infinite loop! Let's avoid generating the random values one-by-one on the fly and comparing them for randomness. We might want to use a list shuffled in a deterministic order. I would recommend another algorithm for making this shuffle more efficiently.

In general, our goal is to:
1. Ensure that selected permutations include the original order -- e.g. `0`, `1`, ..., `defaultGoal.size() - 1`.
2. Generate (a subset of) permutations of `shuffledGoal`
3. Pick as many as `_numPrecomputingThreads` elements from possible permutations.

Reply:
I have thought about this when I was writing the code. My concern was we might lose the chance of finding some better goal priority order(or some new findings) if we do this in a deterministic way. 
And in real application scenario, the number of priority combination should easily be much larger than the number of thread we config to do precomputing, so the possibility of a conflict out of randomization will be small and we can generate shuffled pattern in a reasonable time.
I will change the code to generate permutation in deterministic way.